burn-wgpu/src/context/client.rs

```diff
@@ -0,0 +1,162 @@
+use super::WorkGroup;
+use std::sync::Arc;
+use wgpu::{BindGroup, Buffer, ComputePipeline};
+
+#[cfg(feature = "async")]
+pub use async_client::AsyncContextClient;
+#[cfg(not(feature = "async"))]
+pub use sync_client::SyncContextClient;
+
+/// Context client allows to speak with a server to execute tasks on the GPU.
+pub trait ContextClient {
+    /// Copy the source buffer content into the destination buffer.
+    ///
+    /// # Notes
+    ///
+    /// Make sure the source buffer isn't used afterward, since a race condition may happen.
+    ///
+    /// If the source buffer is still used afterward, use [tensor copy](crate::tensor::WgpuTensor::copy)
+    /// instead. This method is still useful to load data from the CPU into a new buffer.
+    fn copy_buffer(
+        &self,
+        buffer_src: Arc<Buffer>,
+        buffer_dest: Arc<Buffer>,
+        wait_for_registered: bool,
+    ) -> Arc<Buffer>;
+    /// Read a [buffer](Buffer).
+    ///
+    /// # Notes
+    ///
+    /// All pending compute tasks will be executed.
+    fn read_buffer(&self, buffer: Arc<Buffer>) -> Vec<u8>;
+    /// Register a new computing task.
+    fn register_compute(
+        &self,
+        bind_group: BindGroup,
+        pipeline: Arc<ComputePipeline>,
+        work_group: WorkGroup,
+    );
+}
+
+#[cfg(feature = "async")]
+mod async_client {
+    use super::ContextClient;
+    use crate::context::{
+        server::{ComputeTask, ContextTask, CopyBufferTask, ReadBufferTask},
+        WorkGroup,
+    };
+    use std::sync::{mpsc, Arc};
+    use wgpu::{BindGroup, Buffer, ComputePipeline};
+
+    /// Client returned by
+    #[derive(new, Debug)]
+    pub struct AsyncContextClient {
+        sender: mpsc::SyncSender<ContextTask>,
+        _server_handle: std::thread::JoinHandle<()>,
+    }
+
+    impl ContextClient for AsyncContextClient {
+        fn copy_buffer(
+            &self,
+            buffer_src: Arc<Buffer>,
+            buffer_dest: Arc<Buffer>,
+            wait_for_registered: bool,
+        ) -> Arc<Buffer> {
+            self.sender
+                .send(CopyBufferTask::new(buffer_src, buffer_dest.clone()).into())
+                .unwrap();
+
+            if !wait_for_registered {
+                return buffer_dest;
+            }
+
+            // Wait for the buffer to be correctly registered so that inplace operations can be
+            // prioritize.
+            loop {
+                std::thread::sleep(std::time::Duration::from_micros(1));
+
+                if Arc::strong_count(&buffer_dest) == 1 {
+                    return buffer_dest;
+                }
+            }
```

Comment:
The exit condition of this loop seems indeterministic. What are the possibilities this exit condition will never hit? Do we want to put a counter if it's a 1,000 (one second), then it panics?

Reply:
I added a assertion before, which will fail before going into the loop if it's called with wrong arguments.

```rust
            if wait_for_registered {
                assert_eq!(Arc::strong_count(&buffer_dest), 1, "You can't wait for the buffer to be registered when multiple references already exist.");
            }
```

It validates that the destination buffer has only one reference, so that it's safe to wait for it to returns to one reference after the cloning to send it to the context server.

Reply:
What if the other party does not drop a reference? We do not have guarantees it will, especially in places like device to device communication (CPU-GPU). Theoretically it'll go infinite and rocket 🚀 will crash =D 

So my advice we program defensively in this case: we count ticks and exist if it takes too long.

Reply:
If the reference isn't dropped it's a bug and it should crash, but I'm not sure if the state is even possible, the only thing that can crash is the channel, and since it's a sync channel, the execution will no go into the while loop. It's really a low level synchonisation between two threads using the atomic int frond the Arc instead of a mutex or another atomic primitive. The goal is to have the lowest performance hit possible while ensuring that inplace operations can be registered on the returned buffer. It should be considered unsafe for performance reason 😅. Otherwise I would have used a channel as callback (similar to how we read a buffer to bytes and sent them back). 

I'll add more documentation about that, but won't implement more assertion, since it may create more bugs or make the loop slower (so the best action would have been to use a callback channel instead).